dair_pll/vis_utils.py

```diff
@@ -121,53 +119,53 @@ def visualize_trajectory(drake_system: DrakeSystem,
 
         max(round(60/11), 1) == 5.
 
-    Code mostly sourced from Greg Izatt (@gizatt) in this Gist::
-
-        https://gist.github.com/gizatt/d96e6b0bcbe6c7ec331efad53a6f11bc
-
     Args:
         drake_system: System associated with provided trajectory.
         x_trajectory: (T, drake_system.space.n_x) state trajectory.
         framerate: desired frames per second of output video.
 
     Returns:
         (1, T, 3, H, W) ndarray video capture of trajectory with resolution
-        H x W, approximately 640 x 480.
-        The true framerate, rounded to an integer
+        H x W, which are set to 480x640 in :py:mod:`dair_pll.drake_utils`.
+        The true framerate, rounded to an integer.
     """
     assert drake_system.plant_diagram.visualizer is not None
     assert x_trajectory.dim() == 2
 
-    # downsample trajectory to approximate framerate.
+    vis = drake_system.plant_diagram.visualizer
+    sim = drake_system.plant_diagram.sim
+
+    # Downsample trajectory to approximate framerate.
     temporal_downsample = max(round((1 / drake_system.dt) / framerate), 1)
     actual_framerate = round((1 / drake_system.dt) / temporal_downsample)
     x_trajectory = x_trajectory[::temporal_downsample, :]
 
-    # Set camera to look at origin from specified point ``camera_position``
-    camera_position = np.array([0., 20., 5.])
-    meshcat_visualizer = drake_system.plant_diagram.visualizer.vis
-    meshcat_visualizer["/Cameras/default/rotated"].set_object(
-        meshcat.geometry.PerspectiveCamera(fov=CAM_FOV, zoom=0.4))
-    meshcat_visualizer['/Cameras/default'].set_transform(
-        RigidTransform(p=camera_position).GetAsMatrix4())
-
-    # Not sure if this part is necessary
-    meshcat_visualizer['/Cameras/default/rotated/<object>'].set_property(
-        'position', [0, 0, 0])
-
-    # Capture frames
-    frames = []
+    # Simulate the system according to the provided data.
     _, carry = drake_system.sample_initial_condition()
     for x_current in x_trajectory:
         drake_system.preprocess_initial_condition(x_current.unsqueeze(0), carry)
-        frame = meshcat_visualizer.get_image()
-        frames.append(frame)
-
-    # composes video ndarray of shape (T, H, W, 4[rgba])
-    video = np.stack([np.array(frame) for frame in frames])
 
-    # remove alpha channel, downsample, and reorder axes to output type
+        # Force publish video frame.
+        sim_context = sim.get_mutable_context()
+        video_context = vis.GetMyContextFromRoot(sim_context)
+        vis._publish(video_context)
+
+    # Compose a video ndarray of shape (T, H, W, 4[rgba]).
+    video = np.stack([np.asarray(frame) for frame in vis._pil_images])
+    vis.Save()
+
+    # Since Drake's VideoWriter defaults to not looping gifs, re-load and re-
+    # save the gif to ensure it loops.  This gif is only for debugging purposes,
+    # as the gif gets overwritten with every trajectory.  The actual output of
+    # this function is a numpy array.
```

Comment:
This made me think -- the reason that this function outputs a numpy array is to use tensorboard's video display function. Now that we're saving as a gif, it's maybe possible that we could instead just give tensorboard the image filename? not sure, but would help with tensorboard versioning issues if this is a change we could make.

Reply:
looking in to this more, not sure it's possible :/